### PR TITLE
Fix settlement overflow on transferred amount

### PR DESCRIPTION
**Pull request by loredanacirstea**
This should fix the overflow issue for the old `deposit + transferred_amount` approach, demonstrated in https://github.com/raiden-network/raiden-contracts/pull/116/files#diff-cf6314cb143087ba6788de239e94298aR7 and mentioned in the thread started here: https://github.com/raiden-network/raiden-contracts/issues/117#issuecomment-395758666 

TODO:
(update: the following is not necessary - detailed explanation in https://github.com/raiden-network/raiden-contracts/pull/123#pullrequestreview-128277844)
- if possible, create test that would trigger an underflow when subtracting the withdrawn amount, to test that we handle it correctly; think some more if this approach is fair.
```
        // Subtract already withdrawn amount
        (participant1_amount, ) = failsafe_subtract(
            participant1_amount,
            participant1_state.withdrawn_amount
        );
```

### CODE REVIEW

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 1]
```diff
@@ -558,58 +576,44 @@ contract TokenNetwork is Utils {
...
-        ) {
-            participant1_amount = 0;
-        }
+        // case2. participant1_transferred_amount can be [0, real_participant1_transferred_amount)
+        // participant1_netted_transferred_amount is bigger than expected and can cause an overflow.
+        // We just take the biggest amount that we can in this case.
+        participant1_amount = failsafe_addition(
+            participant1_netted_transferred_amount,
+            participant1_state.deposit
+        );
```
Comment (loredanacirstea) [inline raiden_contracts/contracts/TokenNetwork.sol]: We can potentially try to re-add the overflowed amount again after subtracting the withdrawn amount. Not entirely sure it is needed. Opinions?

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 2]
```diff
@@ -544,6 +548,20 @@ contract TokenNetwork is Utils {
...
+        // participant2_transferred_amount can be [0, real_participant2_transferred_amount)
+        // We need to punish participant1.
+        // case2. If participant2 does NOT provide a balance proof or provides an old balance proof.
+        // participant1_transferred_amount can be [0, real_participant1_transferred_amount)
+        // We need to punish participant2.
+        // case3. If neither participants provide a balance proof, we just subtract their
+        // withdrawn amounts from their deposits.
+
+        // case1 is solved by making sure we only have to handle case2 or case3:
+        require(participant2_transferred_amount >= participant1_transferred_amount);
```
Comment (err508) [inline raiden_contracts/contracts/TokenNetwork.sol]: This is already done in line 460

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 3]
```diff
@@ -454,6 +456,8 @@ contract TokenNetwork is Utils {
     )
         public
     {
+        // Convention required in getSettleTransferAmounts
+        require(participant2_transferred_amount >= participant1_transferred_amount);
```
Comment (err508) [inline raiden_contracts/contracts/TokenNetwork.sol]: I think you could interchange `participant1` and `participant2` here appropriately if needed, so we have the assumption, but transactions do not fail.

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 4]
```diff
@@ -1160,4 +1164,34 @@ contract TokenNetwork is Utils {
...
+    /// @param a Addend
+    /// @param b Addend
+    /// @return Maximum between the result of the addition or the maximum uint256 value.
+    function failsafe_addition(uint256 a, uint256 b)
+        pure
+        internal
+        returns (uint256)
+    {
+        uint256 sum = a + b;
+        if (sum >= a && sum >= b) {
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: Yes, you are right.

File: raiden_contracts/contracts/TokenNetwork.sol (hunk 2 above)
Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: I really want this check to stay with the `getSettleTransferAmounts` function, to protect from people changing the code without knowing the importance of it.

I would rather remove the one from 460. 
But I also wanted:
- the main `settleChannel` to be aware that the order is important (I can just add a comment in the docstring to solve this one)
- fail faster and return more gas to the sender. (not really an issue)

So, I am fine with adding the info the docstring + removing line 460.

Comment (err508) [reply raiden_contracts/contracts/TokenNetwork.sol]: You could add a comment in order _to protect from people changing the code without knowing the importance of it_ and get the other two for free.

File: raiden_contracts/contracts/TokenNetwork.sol (hunk 1 above)
Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: If we have valid balance proofs, `participant1_netted_transferred_amount + participant1_state.deposit` should never overflow, because we already ensure that `participant1_state.deposit + participant2_state.deposit` does not overflow.

If we do not have valid balance proofs, then it is fair to cap  `participant1_netted_transferred_amount + participant1_state.deposit` at `MAX_UINT256`.

So, I think the current approach is ok. We don't need to do what I proposed in my comment above.

Comment (err508) [reply raiden_contracts/contracts/TokenNetwork.sol]: I think `participant1_state.deposit + participant2_state.deposit < MAX_UINT256 -> participant1_netted_transferred_amount + participant1_state.deposit < MAX_UINT256` does not hold, or do we have a require somewhere that I missed? 
I would, however, also do not add logic to deal with the case we're it might, as it is imo pretty unlikely to occur in a non-adverse scenario.

Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: > does not hold

why? 
I was talking about the case where balance proofs are computed valid. Therefore, if a participant needs to receive a **positive** netted amount, this amount can only come from the partner's `deposit` => `participant1_netted_transferred_amount` <= `participant2_state.deposit`

Comment (err508) [reply raiden_contracts/contracts/TokenNetwork.sol]: participant1_netted_transferred_amount = 2**255 <= 2**255 = participant2_state.deposit would be a valid variable assignment for a balance proof, that would overflow, no?

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 5]
```diff
@@ -22,6 +22,8 @@ contract TokenNetwork is Utils {
     // Chain ID as specified by EIP155 used in balance proof signatures to avoid replay attacks
     uint256 public chain_id;
 
+    uint256 constant public MAX_SAFE_UINT256 = 115792089237316195423570985008687907853269984665640564039457584007913129639935;
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: I was reluctant to use `2 ** 256 - 1`, even though it works correctly now, because `2 ** 256` is not a valid value for `uint256`. 
I was reluctant to use `0 - 1` in case anything changes in the EVM regarding how underflows are handled.

I thought that having the actual number there is on the safer side. Maybe an overkill to think about future EVM changes, but the current implementation should be fine.

[the remark this answers is not in the document]
Comment (hackaugusto) [reply raiden_contracts/contracts/TokenNetwork.sol]: I asked because `115792089237316195423570985008687907853269984665640564039457584007913129639935` is a number that I don't know by heart, so I had to run `2 ** 256 - 1` to check the value was correct.

but I can see your point, either way is fine to me

[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: I also added a test comparing this number with `2 ** 256 - 1` (python) and with `0 - 1` (Solidity)

[the remark this answers is not in the document]
Comment (LefterisJP) [reply raiden_contracts/contracts/TokenNetwork.sol]: > I was reluctant to use 0 - 1 in case anything changes in the EVM regarding how underflows are handled.

Stuff like that won't change anytime soon. It would break way too many production systems.

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 6]
```diff
@@ -544,6 +550,21 @@ contract TokenNetwork is Utils {
         private
         returns (uint256, uint256, uint256, uint256)
     {
+        // Cases require attention:
```
Comment (LefterisJP) [inline raiden_contracts/contracts/TokenNetwork.sol]: Cases **that** require

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 7]
```diff
@@ -544,6 +550,21 @@ contract TokenNetwork is Utils {
...
+        // participant1_transferred_amount can be [0, real_participant1_transferred_amount)
+        // We need to punish participant2.
+        // case3. If neither participants provide a balance proof, we just subtract their
+        // withdrawn amounts from their deposits.
+
+        // case1 is solved by making sure we only have to handle case2 or case3:
+        // we have the following check in settleChannel:
+        // require(participant2_transferred_amount >= participant1_transferred_amount)
+
+        uint256 participant1_netted_transferred_amount;
```
Comment (LefterisJP) [inline raiden_contracts/contracts/TokenNetwork.sol]: netted -> net   both here and in all below occurences

File: raiden_contracts/contracts/TokenNetwork.sol (hunk 1 above)
Comment (LefterisJP) [reply raiden_contracts/contracts/TokenNetwork.sol]: > So, I think the current approach is ok. We don't need to do what I proposed in my comment above.

Yes it is definitely fair to cap participant1 at UINT256_safe_max at the case where participant2 has cheated.

As a sidenote, the comments here are really nice and explain the edge cases nicely but we should also document those in the spec eventually

File: raiden_contracts/tests/fixtures/channel.py [hunk 8]
```diff
@@ -213,16 +222,11 @@ def get(
...
-            participant1,
-            transferred_amount1,
-            locked_amount1,
-            locksroot1,
-            participant2,
-            transferred_amount2,
-            locked_amount2,
-            locksroot2
-        ).transact({'from': participant1})
+        participant1_values = ChannelBalanceHashData(transferred=transferred_amount1, locked=locked_amount1, locksroot=locksroot1)
```
Comment (LefterisJP) [inline raiden_contracts/tests/fixtures/channel.py]: I think those lines are too big, no?

File: raiden_contracts/tests/test_monitoring_service.py [hunk 9]
```diff
@@ -107,14 +107,14 @@ def test_msc_happy_path(
...
-        0,                   # participant1_locked_amount
-        b'\x00' * 32,        # participant1_locksroot
         B,                   # participant2
         10,                  # participant2_transferred_amount
         0,                   # participant2_locked_amount
         b'\x00' * 32,        # participant2_locksroot
+        A,                   # participant1
+        20,                  # participant1_transferred_amount
+        0,                   # participant1_locked_amount
+        b'\x00' * 32,        # participant1_locksroot
```
Comment (LefterisJP) [inline raiden_contracts/tests/test_monitoring_service.py]: Just use `EMPTY_MERKLE_ROOT` from utils/merkle.py for this and the above. Optionally move it into a more central utils location.

File: raiden_contracts/tests/test_print_gas.py [hunk 10]
```diff
@@ -129,8 +129,8 @@ def test_channel_cycle(
 
     web3.testing.mine(settle_timeout)
     txn_hash = token_network.functions.settleChannel(
+        B, 5, locked_amount2, locksroot2,
```
Comment (LefterisJP) [inline raiden_contracts/tests/test_print_gas.py]: style: each arg in its own line

**Comment by loredanacirstea** [comment]
Opened https://github.com/raiden-network/raiden-contracts/issues/127 for the locked amount attack (old valid balance proof, high locked amount)
